packages/gatsby/src/commands/serve.js

```diff
@@ -1,13 +1,39 @@
 /* @flow weak */
 const path = require(`path`)
 const openurl = require(`better-opn`)
+const fs = require(`fs-extra`)
 const signalExit = require(`signal-exit`)
 const compression = require(`compression`)
 const express = require(`express`)
-const historyFallback = require(`express-history-api-fallback`)
 const getConfigFile = require(`../bootstrap/get-config-file`)
 const preferDefault = require(`../bootstrap/prefer-default`)
 const chalk = require(`chalk`)
+const { match: reachMatch } = require(`@reach/router/lib/utils`)
+
+const getPages = directory =>
+  fs
+    .readFile(path.join(directory, `.cache`, `pages.json`))
+    .then(contents => JSON.parse(contents))
+    .catch(() => [])
+
+const historyRouter = (pages, options) => {
```

Comment:
NIT: maybe we could name it `clientOnlyPathsRouter`/`matchPathRouter` or something more descriptive to what it actually do - I find `historyRouter` name to not match what it actually does

Reply:
Agreed! It's the convention from the existing solution (e.g. https://www.npmjs.com/package/express-history-api-fallback) but yeah--this is nicer.

Reply:
Yeah, I totally understand - I even started writing that I assume it comes from currently used package name (which I also find confusing :P)